internal/util/ssa/patch_test.go

```diff
@@ -33,47 +37,135 @@ func TestPatch(t *testing.T) {
+		// Compute request identifier, so we can later verify that the update call was not cached.
+		modifiedUnstructured, err := prepareModified(env.Scheme(), modifiedObject)
+		g.Expect(err).ToNot(HaveOccurred())
+		requestIdentifier, err := ComputeRequestIdentifier(env.GetScheme(), originalObject, modifiedUnstructured)
+		g.Expect(err).ToNot(HaveOccurred())
+		// Update the object
+		g.Expect(Patch(ctx, env.GetClient(), fieldManager, modifiedObject, WithCachingProxy{Cache: ssaCache, Original: originalObject})).To(Succeed())
+		// Verify that request was not cached (as it changed the object)
+		g.Expect(ssaCache.Has(requestIdentifier)).To(BeFalse())
+
+		// 3. Repeat the same update and verify that the request was cached as the object was not changed.
+		// Get the original object.
+		originalObject = initialObject.DeepCopy()
+		g.Expect(env.GetAPIReader().Get(ctx, client.ObjectKeyFromObject(originalObject), originalObject))
+		// Modify the object
+		modifiedObject = initialObject.DeepCopy()
+		g.Expect(unstructured.SetNestedField(modifiedObject.Object, "baz", "spec", "foo")).To(Succeed())
+		// Compute request identifier, so we can later verify that the update call was cached.
+		modifiedUnstructured, err = prepareModified(env.Scheme(), modifiedObject)
+		g.Expect(err).ToNot(HaveOccurred())
+		requestIdentifier, err = ComputeRequestIdentifier(env.GetScheme(), originalObject, modifiedUnstructured)
+		g.Expect(err).ToNot(HaveOccurred())
+		// Update the object
+		g.Expect(Patch(ctx, env.GetClient(), fieldManager, modifiedObject, WithCachingProxy{Cache: ssaCache, Original: originalObject})).To(Succeed())
+		// Verify that request was cached (as it did not change the object)
+		g.Expect(ssaCache.Has(requestIdentifier)).To(BeTrue())
+	})
+
+	t.Run("Test patch with Machine", func(t *testing.T) {
+		// Build the test object to work with.
+		initialObject := &clusterv1.Machine{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: clusterv1.GroupVersion.String(),
+				Kind:       "Machine",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "machine-1",
+				Namespace: ns.Name,
+				Labels: map[string]string{
+					"label": "labelValue",
+				},
+				Annotations: map[string]string{
+					"annotation": "annotationValue",
+				},
+			},
+			Spec: clusterv1.MachineSpec{
+				ClusterName:      "cluster-1",
+				Version:          pointer.String("v1.25.0"),
+				NodeDrainTimeout: &metav1.Duration{Duration: 10 * time.Second},
+				Bootstrap: clusterv1.Bootstrap{
+					DataSecretName: pointer.String("data-secret"),
+				},
+			},
+		}
+		fieldManager := "test-manager"
+		ssaCache := NewCache()
+
+		// 1. Create the object
+		createObject := initialObject.DeepCopy()
+		g.Expect(Patch(ctx, env.GetClient(), fieldManager, createObject)).To(Succeed())
+		// Note: We have to patch the status here to explicitly set these two status fields.
+		// If we don't do it the Machine defaulting webhook will try to set the two fields to false.
```

Comment:
Out of curiosity: This is because these two fields are type `bool` and do not have `"omitempty"`, right?

Reply:
Yes. Because they have omitempty our mutating/defaulting webhook sees a diff after marshaling old / new object (i.e. old object doesn't have the bools new ones has it).

Basically on initial create status is not written at all (because I'm just using the normal client not the status client). Then the webhook when comparing old/new after marshalling and when calculating the patches it should return always sees the diff. The webhook is sending back the patches to modify the status every time it is called.

For some reason on the 2nd update call the apiserver actually applies those patches (and thus bumps the resourceVersion).